cmd/server/command/migrate.go

```diff
@@ -0,0 +1,304 @@
+					}
+				}
+			}
+			gitSvc := git.Service{
+				Tracer:            tracer,
+				SSHPrivateKeyPath: startOptions.configRepo.SSHPrivateKeyPath,
+				ConfigRepoURL:     startOptions.configRepo.ConfigRepo,
+				Config:            startOptions.gitConfigOpts,
+				ArtifactFileName:  startOptions.configRepo.ArtifactFileName,
+			}
+
+			s3storageSvc, err := s3storage.New(startOptions.s3storage.S3BucketName, tracer)
+			if err != nil {
+				return fmt.Errorf("failed setting up s3 storage: %w", err)
+			}
+
+			var _ = s3storageSvc
+
+			ctx := context.Background()
+			close, err := gitSvc.InitMasterRepo(ctx)
+			if err != nil {
+				return fmt.Errorf("failed initing master repo: %w", err)
+			}
+			defer close(ctx)
+
+			readDir, closeReadDir, err := git.TempDir(ctx, tracer, "read")
+			if err != nil {
+				return fmt.Errorf("failed creating tmp read dir: %w", err)
+			}
+			defer closeReadDir(ctx)
+			_, err = gitSvc.Clone(ctx, readDir)
+			if err != nil {
+				return fmt.Errorf("failed creating tmp read dir: %w", err)
+			}
+
+			serviceAndNamespace := map[string]string{}                // serviceName -> namespace
+			serviceArtifacts := map[string]map[string]artifact.Spec{} // serviceName -> map of artifact IDs -> artifact.Spec
+			for _, env := range []string{"dev", "staging", "prod"} {
+				namespaceDirs, err := ioutil.ReadDir(path.Join(readDir, env, "releases"))
+				if err != nil {
+					return err
+				}
+				for _, namespaceDir := range namespaceDirs {
+					if !namespaceDir.IsDir() {
+						continue
+					}
+
+					serviceDirs, err := ioutil.ReadDir(path.Join(readDir, env, "releases", namespaceDir.Name()))
+					if err != nil {
+						return err
+					}
+					for _, serviceDir := range serviceDirs {
+						if !serviceDir.IsDir() {
+							continue
+						}
+						releasePath := path.Join(readDir, env, "releases", namespaceDir.Name(), serviceDir.Name(), "artifact.json")
+						artifactSpec, err := artifact.Get(releasePath)
+						if err != nil {
+							return err
+						}
+						serviceAndNamespace[serviceDir.Name()] = namespaceDir.Name()
+						if _, ok := serviceArtifacts[serviceDir.Name()]; !ok {
+							serviceArtifacts[serviceDir.Name()] = make(map[string]artifact.Spec)
+						}
+						serviceArtifacts[serviceDir.Name()][artifactSpec.ID] = artifactSpec
+					}
+				}
+			}
+
+			var artifactResult []ArtifactInfo
+			for service, artifactsMap := range serviceArtifacts {
+				for artifactID, artifactSpec := range artifactsMap {
+					exists, err := s3storageSvc.ArtifactExists(ctx, service, artifactID)
+					if err != nil {
+						return err
+					}
+					if exists {
+						artifactResult = append(artifactResult, ArtifactInfo{
+							ServiceName: service,
+							ArtifactID:  artifactID,
+							ExistsInS3:  true,
+							Spec:        artifactSpec,
+						})
+						continue
+					}
+
+					specPath, _, artifactClose, err := gitSvc.ArtifactPaths(ctx, service, "dev", artifactSpec.Application.Branch, artifactID)
```

Comment:
This might be a slow point as `ArtifactPaths` copy the master repo, but lets see how it goes :)